### PR TITLE
Parse OSM node IDs // create unique stable edge IDs

### PR DESCRIPTION
**Pull request by mark-idleman**
https://replicahq.atlassian.net/browse/RAD-4946

This PR broadly hits 2 goals:
- Parse, store, and output OSM node ID information in our street exports for every edge
- Make stable edge IDs _actually_ unique by utilizing this new node ID info

The "meat" of the PR is in the new `CustomOsmReader` class, that overrides some core functions that Graphhopper uses when initially parsing OSM. Every time a graphhopper edge is built within the reader, we now store the OSM node IDs associated with the start/end vertices of that edge (`node_mapping`). In addition, we store mappings of `<ID of new graphhopper edge> -> <ID of base graphhopper node associated with the edge>` (`edge_base_mapping`) and `<ID of new graphhopper edge> -> <ID of adjacent graphhopper node associated with the edge>` (`edge_adjacent_mapping`); this is because later on, when we're building the street export by looping through every internal graphhopper edge, I've found that the graphhopper nodes associated with those edges don't line up with the graphhopper node IDs we stored during the OSM reader phase (I'm guessing some other post-processing step has modified these GH node IDs somehow). I also switched our old logic for storing OSM Way IDs for every edge to use this new pattern as well (stored in `edge_mapping`).

This means that during the street export, our process for exporting the proper OSM node IDs for each graphhopper edge is:
1. Find the "original" graphhopper node IDs that were associated with the edge when it was first created by keying into `edge_base_mapping` and `edge_adjacent_mapping` with the edge's ID
2. With the "original" start/end graphhopper node IDs found from step 1, key into the `node_mapping` map to find the OSM node ID associated with each graphhopper node

---
With this new OSM node ID info, we're able to update the stable edge ID hash calculation to ensure _actually_ unique IDs. The new calculation is as follows:
```
MD5("<startOsmNodeId> <endOsmNodeId> <osmWayId> <geometryString>")
```
The geometry string is needed because of edges that make up loops - they often will have identical way + start/end node IDs, but their geoms will differ in direction, meaning geoms are the only differentiating factor between these types of edges.

---
How does this line up against Inrix's export? It seems to line up quite well, except for a few edge-cases we haven't addressed fully here:
1. OSM Ways that contain [barrier nodes](https://wiki.openstreetmap.org/wiki/Key:barrier). Take a Way with nodes `A -> B -> C`, where `B` is a barrier node. Inrix produces one link, `A -> C`, while we produce 2, `A -> B` and `B -> C`.
2. Certain loop edges. Take a loop edge `A -> B -> A` with osm Way ID `X`. We're essentially producing `A -> B` and `B -> A` (with each link's geoms representing which part of the loop the edges come from), while Inrix seems to have a real bug in their code; they'll produce something like `A -> B` and `X -> A` (yes, they insert the _way_ ID as a _node_ ID 🤔 )
3. OSM Ways that are intersected by walking paths/sidewalks/etc. For a Way `A -> B -> C` where `B` is a node that's intersecting with a foot path, we produce `A -> B` and `B -> C` while Inrix produces just `A -> C`. My guess is, because Inrix only cares about car-centric volumes/network elements, they might only be using the graphhopper auto encoders when constructing their graph, which may ignore non-car-accessible OSM elements altogether. We obviously do care about walking routing, so this isn't an option for us, and these intersections are in fact "meaningful" in a Replica context

---
For a nationwide table I generated with this code, we saw 134,801,141 distinct stable edge IDs out of 134,801,839 total rows. Spot-checking the few non-unique stable edge IDs that remain, it turns out the entire rows they're part of are identical, so we can simply unique these rows away before/during our upload of the exports to BQ.

---
Inrix matching quality of this new scheme is overviewed in slack [here](https://replicahq.slack.com/archives/C03AMSR3SCA/p1655496950190059)

### CODE REVIEW

File: web-bundle/src/main/java/com/graphhopper/replica/StreetEdgeExporter.java [hunk 1]
```diff
@@ -98,17 +103,34 @@ public List<StreetEdgeExportRecord> generateRecords(EdgeIteratorState iteratorSt
...
+            return output;
+        }
+
+        // Fetch OSM Node IDs for each node of edge
+        long startOsmNode = osmHelper.getOSMNode(osmHelper.getBaseNodeForEdge(ghEdgeId));
+        long endOsmNode = osmHelper.getOSMNode(osmHelper.getNodeAdjacentToEdge(ghEdgeId));
+
+        // Filter out single-point "edges" + edges with identical start/end point locations
+        // and no intermediate points (as would exist in the case of road loops)
+        if (wayGeometry.size() <= 1) {
```
Comment (mark-idleman) [inline web-bundle/src/main/java/com/graphhopper/replica/StreetEdgeExporter.java:118]: Special callout here - I added some extra cases where we filter out edges in the export, mostly based on trial-and-error and finding weird edge cases where we definitely don't want "silly" links included in the export

File: web/src/test/java/com/replica/StreetEdgeExporterTest.java [hunk 2]
```diff
@@ -34,7 +32,23 @@ public void testExportEndToEnd() throws IOException {
         File expectedOutputLocation = new File(EXPORT_FILES_DIR + "street_edges.csv");
         CSVParser parser = CSVParser.parse(expectedOutputLocation, StandardCharsets.UTF_8, format);
         List<CSVRecord> records = parser.getRecords();
-        assertEquals(1102288, records.size());
+        assertEquals(1095756, records.size());
+
+        // Sanity check OSM node + edge coverage and stable edge ID uniqueness
```
Comment (mark-idleman) [inline web/src/test/java/com/replica/StreetEdgeExporterTest.java:37]: New tests to keep us feeling confident :)

File: reader/src/main/java/com/graphhopper/reader/osm/CustomOsmReader.java [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+        edgeBaseMapping.setInt(pointer, bitUtil.getIntLow(adjacentNodeId));
+        edgeBaseMapping.setInt(pointer + 4, bitUtil.getIntHigh(adjacentNodeId));
+    }
+
+    /*
+    Copy of OSMReader.addOsmWay() that collects additional information about OSM node IDs
+    and the base/adjacent graphhopper nodes associated with each new graphhopper edge
+    */
+    @Override
+    Collection<EdgeIteratorState> addOSMWay(final LongIndexedContainer osmNodeIds, final IntsRef flags, final long wayOsmId) {
```
Comment (mark-idleman) [inline reader/src/main/java/com/graphhopper/reader/osm/CustomOsmReader.java:70]: This function was basically copied/pasted from the core graphhopper code - the only difference is the instances where we're calling the functions above this one, used for storing node IDs + base/adj node info for each edge

**Review by kaelgreco — APPROVED** [review]

**Comment by mark-idleman** [comment]
Current plan, for posterity: Merge this PR, feeling decent about the testing we've done with it up to this point. We have a "scheduled" opportunity to build a complete region coming up very soon, so we can use that as a chance to double check that these changes didn't cause any issues in prod.

Our deadline for starting the next set of full regional builds has been pushed back, so the focus is now to get a graphhopper core update done + into testing ASAP, with the hope of using that in the upcoming runs. I'll aim to essentially reimplement the features of this PR along with those chances, as it seems like a lot has been changed and this PR might not be relevant there. What I've learned here should make those changes a lot more straightforward, though
